hf_olmo/convert_olmo_to_hf.py

```diff
@@ -60,46 +101,154 @@ def write_tokenizer(checkpoint_dir: str):
+            local_file_path = os.path.join(local_dir, os.path.relpath(s3_key, prefix))
+            local_file_dir = os.path.dirname(local_file_path)
+
+            # Ensure local directory exists
+            if not os.path.exists(local_file_dir):
+                os.makedirs(local_file_dir)
+
+            # Download the file
+            s3_client.download_file(bucket_name, s3_key, local_file_path)
+
+            # Update the progress bar
+            pbar.update(1)
+
+
+@contextmanager
+def make_local_checkpoint(checkpoint_dir: str) -> Generator[str, None, None]:
+    parsed_dir = urlparse(checkpoint_dir)
+
+    assert parsed_dir.scheme in ["s3", ""], "Only s3 and local paths are supported."
+
+    if os.path.exists(checkpoint_dir):
+        yield checkpoint_dir
+        return
+
+    temp_dir = os.path.join(tempfile.gettempdir(), md5(checkpoint_dir.encode()).hexdigest())
+    if os.path.exists(temp_dir):
+        yield temp_dir
+        return
+    try:
+        os.makedirs(temp_dir, exist_ok=True)
+        print(f"Downloading checkpoint to {temp_dir}...")
+        download_s3_directory(
+            bucket_name=parsed_dir.netloc,
+            prefix=parsed_dir.path[1:],
+            local_dir=temp_dir,
+            ignore=r"/(optim|train)/",
+        )
+    except Exception as e:
+        logger.error(f"Error downloading checkpoint: {e}")
+        shutil.rmtree(temp_dir)
+        raise e
+
+    yield temp_dir
+
+
+@contextmanager
+def upload_local_checkpoint(local_checkpoint_dir: str, destination_dir: str) -> Generator[None, None, None]:
+    yield
+
+    if destination_dir == local_checkpoint_dir:
+        return
+    elif (parsed_url := urlparse(destination_dir)).scheme == "s3":
+        s3_bucket_name = parsed_url.netloc
+        s3_prefix = parsed_url.path[1:]
+
+        local_paths = [
+            os.path.join(root, post_fn)
+            for root, _, files in os.walk(local_checkpoint_dir)
+            for post_fn in files
+            if os.path.basename(post_fn) in HF_FILENAMES
+        ]
+        dest_paths = [
+            os.path.join(s3_prefix, os.path.relpath(local_path, local_checkpoint_dir))
+            for local_path in local_paths
+        ]
+
+        s3_client = boto3.client("s3")
+        for local_path, dest_path in tqdm(
+            zip(local_paths, dest_paths), desc="Uploading files", total=len(local_paths)
+        ):
+            s3_client.upload_file(local_path, s3_bucket_name, dest_path)
+    elif parsed_url.scheme == "":
+        shutil.copytree(local_checkpoint_dir, destination_dir)
+    else:
+        raise ValueError(f"Unsupported destination: {destination_dir}. Only s3 and local paths are supported.")
+
+
+def maybe_unshard(checkpoint_dir: str):
+    if os.path.exists(os.path.join(checkpoint_dir, "model.pt")):
+        return
+
+    print(f"Unsharding {checkpoint_dir}...")
+    train_config = TrainConfig.load(os.path.join(checkpoint_dir, "config.yaml"))
+    checkpointer = OlmoCoreCheckpointer(train_config)
```

Comment:
```suggestion
    checkpointer = build_sharded_checkpointer(train_config)
```
This would allow other types of sharded checkpoints to work here.